plugins/se/common.h

```diff
@@ -0,0 +1,87 @@
+// Copyright 2019 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+#pragma once
+
+typedef std::map<std::string, procptr_t> Interfaces;
+
+struct InterfaceReg {
+	uint32_t createFunction;
+	uint32_t name;
+	uint32_t next;
+};
+
+static Interfaces getInterfaces(const procptr_t &module) {
+	Interfaces interfaces;
+
+	// s_pInterfaceRegs is exported on Linux
+	auto s_pInterfaceRegs = getExportedSymbol("s_pInterfaceRegs", module);
+	if (!s_pInterfaceRegs) {
+		const auto CreateInterface = getExportedSymbol("CreateInterface", module);
+		if (CreateInterface == 0) {
+			return interfaces;
+		}
+
+		bool jmpOnly;
+
+		if (peekProc<uint8_t>(CreateInterface) == 0xE9) {
+			// Left 4 Dead:
+			// E9 ?? ?? ?? ??    jmp    CreateInterface_0
+			jmpOnly = true;
+		} else {
+			// Other games:
+			// 55                push    ebp
+			// 8B EC             mov     ebp, esp
+			// 5D                pop     ebp
+			// E9 ?? ?? ?? ??    jmp     CreateInterfaceInternal
+			jmpOnly = false;
+		}
+
+		const auto jmpTarget = peekProc<int32_t>(CreateInterface + (jmpOnly ? 1 : 5));
+		const auto jmpInstructionEnd = CreateInterface + (jmpOnly ? 5 : 9);
+		const auto CreateInterfaceInternal = jmpInstructionEnd + jmpTarget;
+
+		// Left 4 Dead:
+		// 56                   push    esi
+		// 8B 35 ?? ?? ?? ??    mov     esi, s_pInterfaceRegs
+		// 85 F6                test    esi, esi
+		// 57                   push    edi
+		//
+		// Other games:
+		// 55                   push    ebp
+		// 8B EC                mov     ebp, esp
+		// 56                   push    esi
+		// 8B 35 ?? ?? ?? ??    mov     esi, s_pInterfaceRegs
+		if (peekProc<uint16_t>(CreateInterfaceInternal + (jmpOnly ? 1 : 4)) != 0x358B) {
+			return interfaces;
+		}
+
+		s_pInterfaceRegs = peekProc<uint32_t>(CreateInterfaceInternal + (jmpOnly ? 3 : 6));
+	}
+
+	auto iface = peekProc<InterfaceReg>(peekProcPtr(s_pInterfaceRegs));
+ 
+	while (true) {
+		const auto name = peekProcString(iface.name);
+		const auto address = peekProc<uint32_t>(iface.createFunction + (isWin32 ? 1 : 2));
+
+		interfaces.insert(Interfaces::value_type(name, address));
+ 
+		if (!iface.next || !peekProc(iface.next, iface)) {
+			break;
+		}
+	}
```

Comment:
Imo this should be replaced with a do-while-construct (see https://stackoverflow.com/questions/390481/is-while-true-with-break-bad-programming-practice):

```
    do {
		const auto name = peekProcString(iface.name);
		const auto address = peekProc<uint32_t>(iface.createFunction + (isWin32 ? 1 : 2));

		interfaces.insert(Interfaces::value_type(name, address));
	} while (!iface.next || !peekProc(iface.next, iface));
```

That way we get rid of the `while (true)` statment which gives me the shivers every time I read that in a code-base xD

Reply:
I agree.

If I recall correctly I used `while (true)` while testing and then forgot to change it to a do-while loop.